### ISSUE TITLE
Does Firebase Admin SDK support Firestore Emulator?

### ISSUE THREAD

**Issue by ChromeQ** [firebase/firebase-admin-node#575]
I have a project using firestore, functions and hosting.
Until recently I was happily working locally with `firebase serve --only functions,hosting` and that would work with the production firestore.
Using the new firestore emulator I read loads suggesting to use emulator over serve, and my functions have firestore triggered functions which I would like to test locally.

When I `firebase serve --only functions` I get a `Ignoring trigger "onDocumentCreate" because the Cloud Firestore emulator is not running.` This suggests the trigger can work with the emulator.
When I have the firestore emulator running in another process, the functions serve does not detect the firestore emulator running.
So finally I try `firebase emulators:start --only firestore,functions` which starts up fine with all triggers.

As soon as Firestore code is executed I get the error:
```
Google API requested!
    - URL: "https://oauth2.googleapis.com/token"
    - Be careful, this may be a production service.
>  { Error: Getting metadata from plugin failed with error: Header field "authorization" must have only a single value
>      at Http2CallStream.call.on (/home/dav/dev/lms/functions/node_modules/@grpc/grpc-js/build/src/call.js:68:41)
>      at emitOne (events.js:121:20)
>      at Http2CallStream.emit (events.js:211:7)
>      at process.nextTick (/home/dav/dev/lms/functions/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
>      at _combinedTickCallback (internal/process/next_tick.js:132:7)
>      at process._tickCallback (internal/process/next_tick.js:181:9)
>    code: 'ERR_HTTP2_HEADER_SINGLE_VALUE',
>    details: 'Getting metadata from plugin failed with error: Header field "authorization" must have only a single value',
>    metadata: Metadata { options: undefined, internalRepr: Map {} } }
```

How can I make firebase-admin work with emulated firestore and functions?
```
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';

admin.initializeApp();
const firestore = admin.firestore();

const foo = firestore.collection('colName').doc('docId')
    .onSnapshot(snap => {
        console.log(snap);
    });
```

Using all the latest packages:
```    "firebase": "^6.2.0",
    "firebase-admin": "^8.1.0",
    "firebase-functions": "^3.0.1",
    "firebase-tools": "^7.0.0",
```
OS: Linux Ubuntu 18.04
NPM version: 6.9.0
Node version: 8.15.0

**Comment by schmidt-sebastian** [comment]
@ChromeQ The easiest way to achieve this to use the recent version of `@google-cloud/firestore`. Starting with version 2.2, we support the environment variable `FIRESTORE_EMULATOR_HOST`. If you set this environment variable to the host entry provided at emulator startup (such as `localhost:8080`), Firestore will connect to the emulator instead of to the backend.

**Comment by ChromeQ** [comment]
Thanks, but that also didn't work as I get another problem:
`npx firebase emulators:start --only firestore` runs in one process and shows `FIRESTORE_EMULATOR_HOST=localhost:5002`

Then in a new process I run:
`FIRESTORE_EMULATOR_HOST=localhost:5002 npx firebase serve --only functions`

```
Your code has been provided a "firebase-admin" instance.
⚠  Error: Cannot set both "settings.host" and "settings.servicePath".
    at Firestore.validateAndApplySettings (/home/dav/dev/lms/functions/node_modules/@google-cloud/firestore/build/src/index.js:353:23)
    at Firestore.settings (/home/dav/dev/lms/functions/node_modules/@google-cloud/firestore/build/src/index.js:339:14)
    at initializeSettings (/home/dav/dev/lms/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:238:33)
    at Proxy.<anonymous> (/home/dav/dev/lms/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:269:21)
    at Object.<anonymous> (/home/dav/dev/lms/functions/index.js:43:11)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
⚠  Your function was killed because it raised an unhandled error.
```

**Comment by schmidt-sebastian** [comment]
We will get that fixed! Sorry for the trouble.

**Comment by ChromeQ** [comment]
Thanks, but ideally it would work with simply: `firebase emulators:start --only firestore,functions` or `firebase serve --only firestore,functions`
Can it be fixed to work like that without the need of the extra variable or extra process running firestore emulator?
Thanks

**Comment by hiranya911** [comment]
@schmidt-sebastian where did we land with this one?

**Comment by schmidt-sebastian** [comment]
Waiting on https://github.com/googleapis/nodejs-firestore/pull/696, which will fix the issue in https://github.com/firebase/firebase-admin-node/issues/575#issuecomment-503673385

**Comment by ChromeQ** [comment]
Just out of curiousity, what is the difference between firebase serve and firebase emulators:start?
I'm only switching over to using emulators:start cos it allows me to config a port in the firebase.json since my port 8080 is used

**Comment by fastfedora** [comment]
FYI, I'm getting the same `ERR_HTTP2_HEADER_SINGLE_VALUE` when executing Firestore code using `firebase emulators:start --only functions`. 

In my case, I'm not trying to emulate the hosting or the Firestore database, only the functions. This previously worked for me in v6.5.0.

**Comment by schmidt-sebastian** [comment]
@ryanpbrewster Ryan, can you look at these (comment https://github.com/firebase/firebase-admin-node/issues/575#issuecomment-504528452 and https://github.com/firebase/firebase-admin-node/issues/575#issuecomment-505003527)?

**Comment by ryanpbrewster** [comment]
`firebase serve` is an old-school command that's usually used to emulate Firebase Hosting (i.e., serve your static web assets).

`firebase serve --only firestore` additionally starts the Firestore emulator. It is very simple and does not integrate well with other emulators or provide useful logging.

`firebase emulators:start` is the recommended way of running the Firebase emulators (including the Firestore emulator). It starts a suite of interacting emulators, and handles logging and configuration.

**Comment by ryanpbrewster** [comment]
`ERR_HTTP2_HEADER_SINGLE_VALUE` indicates that somewhere we're setting a header more than once. From the original report it seems like this is an issue with the underlying `@google-cloud/firestore` library which will be fixed once https://github.com/googleapis/nodejs-firestore/pull/696 is merged and released.

One possible workaround, for now, is to downgrade to `@google-cloud/firestore@2.0.0`

**Comment by ChromeQ** [comment]
Thanks for the explanation @ryanpbrewster.
That's kind of what I thought, so will it be possible to run the firestore and functions emulators as I asked in https://github.com/firebase/firebase-admin-node/issues/575#issuecomment-503673385

**Comment by ryanpbrewster** [comment]
Unfortunately no. `firebase emulators:start --only functions,firestore` and `firebase serve --only functions,firestore` both actually start the Firestore emulator (in addition to the Functions emulator). This is very hard to work around. If you want to test functions that are triggered by Firestore operations, you'll need to run the Firestore emulator.

**Comment by ChromeQ** [comment]
I have updated to the latest version (v2.2.3 of @google-cloud/firestore) and can confirm that this is working as expected, I can now run `firebase emulators:serve --only firestore,functions` and it will use the expected ports from `firebase.json` and both emulators work together without errors. Thanks @ryanpbrewster & @schmidt-sebastian

**Comment by hiranya911** [comment]
@schmidt-sebastian can this be closed? Is there any action items for the firebase-admin package?

**Comment by lookfirst** [comment]
Can someone provide a detailed explanation of how to get this running? The documentation is missing imho. I found this project https://github.com/firebase/quickstart-nodejs/tree/master/firestore-emulator/typescript-quickstart

But that uses the firestore js sdk and not the admin sdk.

So, what exactly do I need to do pass to initializeApp for the admin sdk so that it hits the local emulator?

**Comment by lookfirst** [comment]
Ok, I finally figured this out, I wish someone at google would document this better because in the end, it isn't very hard to setup... 

Start the emulator:

`firebase emulators:start --only firestore`

```ts
import {clearFirestoreData} from "@firebase/testing";
import * as admin from "firebase-admin";

process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';

const app = admin.initializeApp({
	projectId: 'test',
	credential: admin.credential.applicationDefault()
});

describe("test", () => {
	beforeEach("reset the database", async () => {
		await clearFirestoreData({projectId: app.options.projectId || ''});
	});

	it("works", async () => {
            // write some tests against the admin sdk firestore
           app.firestore()....
	});
});
```

**Comment by Elyx0** [comment]
I'm hitting the same problem here
with `firebase emulators:start --only functions,database,firestore`
```
'use strict';
import 'jest';

process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
process.env.FIREBASE_DATABASE_EMULATOR_HOST = 'localhost:9000';

// INIT ---------------------------------------------------------------------------------- 
import * as functions from 'firebase-functions';
import * as functionsTest from 'firebase-functions-test';
import * as admin from 'firebase-admin';
admin.initializeApp();
```
But then any code calling `admin` throws with `Can't determine Firebase Database URL.`
I run my tests with
```
FIREBASE_DATABASE_EMULATOR_HOST=localhost:9000 FIRESTORE_EMULATOR_HOST=localhost:8080 jest --watchAll --coverage
```

I see from https://github.com/firebase/firebase-admin-node/commit/a4ad8288d1004b9c801dd7ee1c6b896fd7112ca8 that one of the env override got removed, am I missing something?

**Comment by Elyx0** [comment]
With `"@firebase/testing";` it sets it up correctly 
```
admin.initializeTestApp({ projectId, databaseName, auth })
admin.initializeAdminApp({ projectId, databaseName, auth })
```
The problem is that most of my code relies on 
```
import * as admin from 'firebase-admin';
```
in each file for doing the database management, so I can't mock it

Should I add a wrapper around my `admin` that exports the one from `firebase/testing` if in test mode to my functions and another to live mode from the `firebase-admin` ? It feels contrived since I have a correct emulator running...

**Comment by Elyx0** [comment]
I ended up doing something like that:
```
import * as admin from 'firebase-admin';
import * as adminTesting from "@firebase/testing";


const databaseName = "database-emulator-example";
const projectId = "firestore-emulator-example";
export const coverageUrl = `http://localhost:9000/.inspect/coverage?ns=${databaseName}`;

export type finalExportUnionType = typeof admin | firebase.app.App;
export let finalExport: finalExportUnionType = admin;
if (process.env.NODE_ENV == 'test') {
 console.log('From test mode');
 const adminApp = adminTesting.initializeAdminApp({projectId, databaseName});
 finalExport = adminApp;
}
```
But it feels like an ugly hack

**Comment by akauppi** [comment]
Though this is closed, this may help someone:

If you're wanting to use the Admin SDK with the new Emulator UI, make sure you set up `projectId` parameter to match your Firebase project. Otherwise, data gets written somewhere, but isn't visible in the emulator.

**Comment by milan-codes** [comment]
Still facing this issue, even though the FIREBASE_DATABASE_EMULATOR_HOST is set (as it is stated [here](https://firebase.google.com/docs/emulator-suite/connect_rtdb#admin_sdks)), I still can't get Admin SDK to connect to the emulators.

**Comment by schmidt-sebastian** [comment]
@milan-codes Most of the advice here is for the Firestore emulator.

There is currently a known bug with the Database emulator setting: https://github.com/firebase/firebase-js-sdk/issues/3144 
Do you see errors similar to the ones reported there?

**Comment by milan-codes** [comment]
@schmidt-sebastian Thanks for pointing me to the right issue! This issue seems to be similar to mine, but now I had the chance to test my code on a Windows PC. Everything works fine there, it successfully connects to the emulator. I have no idea why it doesn't work on my mac. Should I open a new issue for this?

**Comment by schmidt-sebastian** [comment]
#3144 is merged and should be released this week. I suggest to wait a couple days before trying to repro.

**Comment by changwo** [comment]
> Ok, I finally figured this out, I wish someone at google would document this better because in the end, it isn't very hard to setup...
> 
> Start the emulator:
> 
> `firebase emulators:start --only firestore`
> 
> ```ts
> import {clearFirestoreData} from "@firebase/testing";
> import * as admin from "firebase-admin";
> 
> process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
> 
> const app = admin.initializeApp({
> 	projectId: 'test',
> 	credential: admin.credential.applicationDefault()
> });
> 
> describe("test", () => {
> 	beforeEach("reset the database", async () => {
> 		await clearFirestoreData({projectId: app.options.projectId || ''});
> 	});
> 
> 	it("works", async () => {
>             // write some tests against the admin sdk firestore
>            app.firestore()....
> 	});
> });
> ```

You are a god send, thank you good sir!

**Comment by ir-fuel** [comment]
FYI @firebase/testing has been deprecated

**Comment by lookfirst** [comment]
> FYI @firebase/testing has been deprecated

what's the replacement?

**Comment by schmidt-sebastian** [comment]
@lookfirst https://www.npmjs.com/package/@firebase/rules-unit-testing

**Comment by vladfaust** [comment]
Please answer if current release supports managing multiple tenants with Emulators? I currently get such a error:

```
 Error: An internal error has occurred. Raw server response: "{"error":{"code":404,"message":"Not Found","errors":[{"message":"Not Found","reason":"notFound"}],"status":"NOT_FOUND"}}"
    at FirebaseAuthError.FirebaseError [as constructor] (C:\Users\Vlad\Development\Desklify\functions\node_modules\firebase-admin\lib\utils\error.js:44:28)
```

for any tenant-related activity, such as `app.auth().tenantManager().listTenants()`. 

I tried everything: setting `databaseURL`, `process.env.FIREBASE_AUTH_EMULATOR_HOST = "localhost:9099"` and `process.env.FIRESTORE_EMULATOR_HOST = "localhost:8080"`. Still no luck.

Thanks in advance!

**Comment by SUVOLewis** [comment]
@vladfaust When trying to create a new user under a specific tenant, I get the following:
```
"{"error":{"code":501,"message":"identitytoolkit.projects.tenants.accounts is not implemented in the Auth Emulator.","errors":[{"message":"identitytoolkit.projects.tenants.accounts is not implemented in the Auth Emulator.","reason":"unimplemented"}],"status":"NOT_IMPLEMENTED"}}"
```

I think it's safe to assume that the Google Cloud Identity Platform Tenant System is not yet a part of the Firebase Auth Emulator.
Currently, I'm unaware of any alternatives to this.
